PEP 661: Update (#4923)
python/peps: peps/pep-0661.rst

Reviewer: ```suggestion
are significant, nor which kind of implementation would be good.  The authors
```

Or is it PEP-style to still refer to the author in singular form even if there are multiple?

Author: I would like to leave Tal's voice preserved as much as possible.

Reviewer: ```suggestion
and continuing discussions, the authors have come to the opinion that a simple,
```

PEP: 661
Title: Sentinel Values
Author: Tal Einat <tal@python.org>, Jelle Zijlstra <jelle.zijlstra@gmail.com>, Victorien Plot <contact@vctrn.dev>
Discussions-To: https://discuss.python.org/t/pep-661-sentinel-values/9126

> Status: Deferred

Reviewer 2: Back to Draft and add a Python version?

```suggestion
Status: Draft
```

Author: Good catch, I'm going to aim for 3.15 (making sure your job stays interesting :) ).

Reviewer 2: Good luck!

However, the common implementations, including some in the stdlib, suffer from
several significant drawbacks.

> This PEP proposes adding a built-in utility for defining sentinel values, to

Reviewer: This feels like a nit, but for me "utility" doesn't cover that we are now also adding a C API. I think the fact that the proposal offers a pure Python and C API is signficant enough to be mentioned in the abstract for readers who quickly skim abstracts of recently approved PEPs.

3. It should be simple to define as many distinct sentinel values as needed.
4. The sentinel objects should have a clear and short repr.
5. It should be possible to use clear type signatures for sentinels.
6. The sentinel objects should behave correctly after copying, and module-scope

>    sentinels should preserve identity when pickled and unpickled.

Reviewer: As this diff shows this criterium actually changed during writing of the diff. Perhaps a better wording would be:

```suggestion
6. The sentinel objects should behave correctly after copying, and sentinels
should have predictable behaviour pickled and unpickled.
```

   documentation.

With so many uses in the Python standard library [2]_, it would be useful to
have an implementation available to the standard library, since the stdlib
cannot use implementations of sentinel objects available elsewhere (such as
the ``sentinels`` [5]_ or ``sentinel`` [6]_ PyPI packages).

After researching existing idioms and implementations, and going through many
different possible implementations, the design below was chosen to meet these
criteria while keeping the API and implementation small (see
`Reference Implementation`_).

Specification
=============

> A new built-in callable named ``sentinel`` will be added.

Reviewer 3: What would be the motivation to make it lowercase? Is it to match existing built-ins that are already all lowercase (and perhaps this is even an established convention)?

Author: Yes, all other builtin types are lowercase (other than exception classes).

This is similar to how ``None`` is handled in the existing type system. For
example::

    MISSING = sentinel('MISSING')

    def foo(value: int | MISSING = MISSING) -> int:
        ...

More formally, type checkers should recognize sentinel creations of the form
``NAME = sentinel('NAME')`` as creating a new sentinel object. If the name
passed to ``sentinel()`` does not match the name the object is

> assigned to, type checkers should emit an error.

Reviewer: Does the PEP need to specify behaviour of:

```python
from typing import assert_type

NOT_MISSING = sentinel('MISSING')

def foo(value: int | NOT_MISSING) -> None:
    if value is NOT_MISSING:
        assert_type(value, NOT_MISSING)
    else:
        assert_type(value, int)
```

Or is that something that naturally follows from the rest of the PEP? Or is that up to each type checker to decide?

I'll defer this to you Jelle as I think you have a much better feeling of what is necessary for type checkers to standardize here, but wanted to make sure this is considered.

Author: I assume you're referring to the name mismatch between the variable and the sentinel.

We already say it's a type checker error to have such a mismatch. Behavior after an error is generally undefined.

Reviewer: Fair enough!

            assert_type(value, int)

To support usage in type expressions, the runtime implementation
of sentinel objects should have the ``__or__`` and ``__ror__``
methods, returning :py:class:`typing.Union` objects.

C API
-----

Sentinels can also be useful in C extensions. We propose two new
C API functions::

* ``PyObject *PySentinel_New(const char *name, const char *module_name)`` creates a new sentinel object.
* ``bool PySentinel_Check(PyObject *obj)`` checks if an object is a sentinel.

C code can use the ``==`` operator to check if an object is a
specific sentinel.

Backwards Compatibility
=======================

> Adding a new builtin means that code which currently relies on the bare name

Reviewer: We could consider adding a note here about the fact that `sentinel` is download much less frequently than `sentinels` according to PyPI Stats:

[sentinels](https://pypistats.org/packages/sentinels) vs. [sentinel](https://pypistats.org/packages/sentinel)
5,6 million vs. 0,1 million last month.

That reduces the scope for backwards compatibility issues significantly.

Author: I'm not sure how relevant that is; there's more choice when introducing a library name, but if we're adding a builtin clearly we're not going to name it `sentinels`.

Reference Implementation
========================

A reference implementation is available as a CPython pull request [10]_.  A
previous reference implementation is found in a dedicated GitHub repo [7]_.
A sketch of the intended behavior follows::

    class sentinel:
        """Unique sentinel values."""

        __slots__ = ("__name__", "_module_name")

        def __init_subclass__(cls):
            raise TypeError("type 'sentinel' is not an acceptable base type")

        def __init__(self, name, /):
            if not isinstance(name, str):
                raise TypeError("sentinel name must be a string")
            self.__name__ = name
            self._module_name = sys._getframemodulename(1)
            if self._module_name is None:

>                 self._module_name = __name__

Reviewer 3: As per https://github.com/JelleZijlstra/cpython/pull/4#discussion_r3099255131, I'm not sure this matches the C implementation, as it defaults to `builtins`?

Author: Yeah I think both cases here are questionable, maybe the best bet is to default it to None and make __reduce__ raise if the module is missing.

25% of the voters.

Use a new standard library module
---------------------------------

Earlier drafts proposed adding a ``Sentinel`` class to a new ``sentinels`` or
``sentinellib`` module.  However, adding a new module for a single public
callable is unnecessary, and using a module makes the feature less convenient
than the existing ``object()`` idiom.  The Steering Council also specifically
encouraged making the feature a builtin so that it is at least as easy to use
as ``object()``.

Using the name ``sentinels`` would also conflict with an existing, actively
used PyPI package.  While other module names are possible, making the feature a

> builtin avoids the naming problem entirely.

Reviewer: Does it? Doesn't it make the naming problem more prominent as you explain in the Backwards Compatibility section?

Author: I do think this is accurate. Backwards compatibility is a problem when code that works in 3.x no longer works when run unchanged in 3.x+1. That would be the case if we picked a new library name, because then in the first version that has this PEP, `import sentinels` (or whatever name we choose) will start importing from the standard library instead of the third-party library. But if we're adding a new builtin, existing code that uses the name `sentinel` is unaffected; it just can't easily use the new functionality.

Use a registry of per-module sentinel names
-------------------------------------------

Earlier drafts proposed making sentinel names unique within each module.  Under

> that design, repeated calls such as ``Sentinel("MISSING")`` from the same

Reviewer: ```suggestion
that design, repeated calls such as ``sentinel("MISSING")`` from the same
```

Considering the rest of this section also uses the new lower case version :)

Author: Good call, this is the only remaining place using the uppercase name.

considered worth the added complexity.

Allowing customization of boolean evaluation
--------------------------------------------

Discussions considered allowing sentinels to be explicitly truthy, falsy, or
not convertible to ``bool``.  Some existing third-party sentinels expose falsy
behavior as part of their public API, and several participants argued that
raising in boolean contexts would better enforce identity checks.

This PEP keeps the initial proposal simpler by giving sentinels the default
truthy behavior of ordinary objects and by recommending identity checks.
Custom boolean behavior may be considered later if the added API and typing
complexity is judged worthwhile.

> Using ``typing.Literal`` in type annotations

Reviewer: I think it is relevant to mention that (I think) this was also the recommendation of the Typing Council.

At least you mention it being discussed here:
https://discuss.python.org/t/pep-661-sentinel-values/9126/153

Was there any follow up on this? I can't find it in the thread.

Either way, `pylance` also added support for this which I think shows that there is agreement within the typing communication about the proposed syntax.

Author: I almost forgot about this :)

We ended up posting on the SC tracker: https://github.com/python/steering-council/issues/258#issuecomment-2417524379

I'll link that post in the PEP.

Additional Notes
================

* This PEP and an initial implementation were drafted in a dedicated GitHub
  repo [7]_.  The implementation should be updated to match this simplified

>   proposal.

Reviewer: It is unclear to me whether "the implementation" refer to the implementation on the repo or in stdlib? I'm guessing it is the latter, at which point I think this is redundant? Shouldn't all implementations in the stdlib follow the final proposal of a PEP? 😄

Author: Going to remove this line
